### PR TITLE
Agent: Update weave to 1.9.3

### LINKED ISSUE #1826: Update weave to 1.9.0
https://github.com/weaveworks/weave/releases/tag/latest_release

### PR DESCRIPTION
**Pull request by jnummelin**
fixes #1826

### CODE REVIEW

**Comment by SpComb** [comment]
Had a chance to test the new IPSec encryption for grid trusted subnets?

**Comment by SpComb** [comment]
Agent oddities when restarting the `kontena-agent`. The weave worker runs `weaveexec attach` commands at the same time as re-launching the new weave router:

```
kontena-agent | I, [2017-03-02T09:50:02.348968 #1]  INFO -- Kontena::Workers::WeaveWorker: initialized
kontena-agent | D, [2017-03-02T09:50:02.817707 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec stream: ["--local", "ps"]
kontena-agent | D, [2017-03-02T09:50:02.825089 #1] DEBUG -- Workers::WeaveWorker: Scanned 4 existing containers for potential migration: {"fe5af6db7c10"=>["10.81.128.92/16"], "a746a4256299"=>["10.81.128.5/16"], "ba04a7ced236"=>["10.81.128.28/16"], "653e4574f253"=>["10.81.128.3/16"]}
kontena-agent | I, [2017-03-02T09:50:02.825367 #1]  INFO -- Kontena::Workers::WeaveWorker: attaching network to existing containers
kontena-agent | D, [2017-03-02T09:50:02.831979 #1] DEBUG -- Workers::WeaveWorker: skip start for container=/kontena-agent without overlay_cidr
kontena-agent | I, [2017-03-02T09:50:02.863790 #1]  INFO -- Kontena::NetworkAdapters::Weave: Attach container=fe5af6db7c101a1314bfe68a7f020b11083378caa5a95e4db38580eb32af2559 at cidr=10.81.128.92/16
kontena-agent | I, [2017-03-02T09:50:02.964824 #1]  INFO -- Kontena::NetworkAdapters::Weave: Attach container=a746a42562999f5cbd9984e8aea6403f69f5dce783622f643e7514371b88f443 at cidr=10.81.128.5/16
kontena-agent | I, [2017-03-02T09:50:03.007447 #1]  INFO -- Kontena::NetworkAdapters::Weave: Attach container=ba04a7ced236046c8886fe852e5ffe12d977c3c3b480cc2dee547f0398129b37 at cidr=10.81.128.28/16
kontena-agent | D, [2017-03-02T09:50:03.018868 #1] DEBUG -- Workers::WeaveWorker: skip start for container=/kontena-cadvisor without overlay_cidr
kontena-agent | D, [2017-03-02T09:50:03.033042 #1] DEBUG -- Workers::WeaveWorker: skip start for container=/weave without overlay_cidr
kontena-agent | D, [2017-03-02T09:50:03.041353 #1] DEBUG -- Workers::WeaveWorker: skip start for container=/kontena-etcd without overlay_cidr
kontena-agent | D, [2017-03-02T09:50:03.051336 #1] DEBUG -- Workers::WeaveWorker: skip start for container=/kontena-ipam-plugin without overlay_cidr
kontena-agent | I, [2017-03-02T09:50:03.089705 #1]  INFO -- Kontena::NetworkAdapters::Weave: Attach container=653e4574f253f572f9c584b279ec5d64dc602cac8f7e646ac8de606df7277316 at cidr=10.81.128.3/16
kontena-agent | D, [2017-03-02T09:50:04.301662 #1] DEBUG -- RpcServer: rpc notification: Kontena::Rpc::AgentApi#master_info [{"version"=>"1.2.0.dev"}]
kontena-agent | D, [2017-03-02T09:50:04.301928 #1] DEBUG -- RpcServer: rpc notification: Kontena::Rpc::AgentApi#node_info [{"id"=>"XI4K:NPOL:EQJ4:S4V7:EN3B:DHC5:KZJD:F3U2:PCAN:46EV:IO4A:63S5", "connected"=>true, "created_at"=>"2016-10-27T08:15:31.084Z", "updated_at"=>"2017-03-02T09:02:03.023Z", "last_seen_at"=>"2017-03-02T09:50:04.260Z", "name"=>"core-01", "os"=>"Container Linux by CoreOS 1298.5.0 (Ladybug)", "engine_root_dir"=>"/var/lib/docker", "driver"=>"overlay", "kernel_version"=>"4.9.9-coreos-r1", "labels"=>["test"], "mem_total"=>1045336064, "mem_limit"=>0, "cpus"=>1, "public_ip"=>"82.181.164.219", "private_ip"=>"192.168.66.101", "overlay_ip"=>"10.81.0.1", "agent_version"=>"1.2.0.dev", "docker_version"=>"1.12.6", "peer_ips"=>["192.168.66.102"], "node_number"=>1, "initial_member"=>true, "grid"=>{"id"=>"development", "name"=>"development", "initial_size"=>1, "token"=>"tBsoco4sBClMtIOD/bHeIqS1KYM0kZMG7B960jNWWQXMOxqhOf8W1QgiNJlW9u6B/qS+rfS2ZLpC/ZJsyPNj0A==", "stats"=>{"statsd"=>nil}, "trusted_subnets"=>[], "subnet"=>"10.81.0.0/16", "supernet"=>"10.80.0.0/12"}, "resource_usage"=>{"memory"=>{"total"=>1045336064, "free"=>218517504, "buffers"=>70524928, "cached"=>329064448, "active"=>497704960, "inactive"=>210034688, "used"=>826818560}, "load"=>{"1m"=>0.3388671875, "5m"=>0.3271484375, "15m"=>0.2177734375}, "filesystem"=>[{"name"=>"/var/lib/docker", "free"=>11233509376, "available"=>10350313472, "used"=>5484883968, "total"=>16718393344}], "usage"=>{"container_seconds"=>522}}}]
kontena-agent | I, [2017-03-02T09:50:04.302566 #1]  INFO -- Kontena::Workers::QueueWorker: started to process msg queue
kontena-agent | D, [2017-03-02T09:50:04.302735 #1] DEBUG -- Workers::NodeInfoWorker: publishing node information
kontena-agent | I, [2017-03-02T09:50:04.311053 #1]  INFO -- Kontena::Workers::LogWorker: started log streaming
kontena-agent | D, [2017-03-02T09:50:04.801681 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec ok: ["--local", "attach", "10.81.128.92/16", "--rewrite-hosts", "fe5af6db7c101a1314bfe68a7f020b11083378caa5a95e4db38580eb32af2559"]
kontena-agent | 10.81.128.92
kontena-agent | 
kontena-agent | D, [2017-03-02T09:50:05.191951 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec ok: ["--local", "attach", "10.81.128.5/16", "--rewrite-hosts", "a746a42562999f5cbd9984e8aea6403f69f5dce783622f643e7514371b88f443"]
kontena-agent | 10.81.128.5
kontena-agent | 
kontena-agent | D, [2017-03-02T09:50:06.356523 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec ok: ["--local", "attach", "10.81.128.28/16", "--rewrite-hosts", "ba04a7ced236046c8886fe852e5ffe12d977c3c3b480cc2dee547f0398129b37"]
kontena-agent | weave container is not present. Have you launched it?
kontena-agent | 10.81.128.28
kontena-agent | 
kontena-agent | D, [2017-03-02T09:50:06.923496 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec ok: ["--local", "attach", "10.81.128.3/16", "--rewrite-hosts", "653e4574f253f572f9c584b279ec5d64dc602cac8f7e646ac8de606df7277316"]
kontena-agent | weave container is not present. Have you launched it?
kontena-agent | 10.81.128.3
kontena-agent | 
kontena-agent | D, [2017-03-02T09:50:07.320000 #1] DEBUG -- Workers::WeaveWorker: skip start for container=/weave without overlay_cidr
kontena-agent | D, [2017-03-02T09:50:07.320382 #1] DEBUG -- Workers::LogWorker: /weave/{"maintainer"=>"Weaveworks Inc <help@weave.works>", "org.label-schema.description"=>"Weave Net creates a virtual network that connects Docker containers across multiple hosts and enables their automatic discovery", "org.label-schema.name"=>"Weave Net", "org.label-schema.url"=>"https://weave.works", "org.label-schema.vcs-url"=>"https://github.com/weaveworks/weave", "org.label-schema.vendor"=>"Weaveworks", "works.weave.role"=>"system"}
kontena-agent | D, [2017-03-02T09:50:07.369845 #1] DEBUG -- Workers::ContainerLogWorker: starting to stream logs from /weave
kontena-agent | D, [2017-03-02T09:50:08.523950 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec ok: ["--local", "launch-router", "--ipalloc-range", "", "--dns-domain", "kontena.local", "--password", "<redacted>", "--trusted-subnets", ""]
kontena-agent | 4c6dcaf54984551a6689c603e253c92820d201197781e576f0e6f157b9c1d429
kontena-agent | 
kontena-agent | D, [2017-03-02T09:50:08.858102 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec ok: ["--local", "connect", "--replace", "192.168.66.102"]
kontena-agent | 
kontena-agent | I, [2017-03-02T09:50:08.940095 #1]  INFO -- Kontena::NetworkAdapters::Weave: router connected to peers 192.168.66.102
kontena-agent | I, [2017-03-02T09:50:08.940240 #1]  INFO -- Kontena::NetworkAdapters::Weave: using trusted subnets: 
kontena-agent | D, [2017-03-02T09:50:09.270014 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec ok: ["--local", "expose", "ip:10.81.0.1/16"]
kontena-agent | 10.81.0.1
kontena-agent | 
kontena-agent | I, [2017-03-02T09:50:09.275493 #1]  INFO -- Kontena::NetworkAdapters::Weave: Exposed host node at cidr=10.81.0.1/16
kontena-agent | D, [2017-03-02T09:50:09.535279 #1] DEBUG -- NetworkAdapters::WeaveExecutor: weaveexec stream: ["--local", "ps", "weave:expose"]
kontena-agent | I, [2017-03-02T09:50:09.560476 #1]  INFO -- Kontena::Launchers::IpamPlugin: network started, launching ipam...
kontena-agent | I, [2017-03-02T09:50:09.572280 #1]  INFO -- Kontena::Launchers::IpamPlugin: ipam-plugin is already running
```

Probably not directly related to this PR.

**Comment by SpComb** [comment]
Upgrade to 1.9.2?

#### `weaveexec status`
```
        Version: 1.9.1 (version 1.9.2 available - please upgrade!)
```

**Comment by SpComb** [comment]
From the [Release Notes](https://github.com/weaveworks/weave/releases/v1.9.2)

> When upgrading from a version earlier than 1.9.0, if your host's network interface has a limit on packet size (the "MTU") smaller than 1496 bytes, you should reboot after upgrading to ensure encrypted fast datapath can work. For instance this applies to Google Cloud Platform, but is not necessary on AWS.

> Weave Net 1.9.2 is fully compatible with all previous versions back to 1.1.0, so it is possible to upgrade clusters incrementally and run mixed versions.

**Comment by SpComb** [comment]
Looks like weave is using `encrypted fastdp` now by default:

```
core@core-02 ~ $ docker run --rm --net host --pid host --privileged -v /var/run/docker.sock:/var/run/docker.sock weaveworks/weaveexec:1.9.2 status            

        Version: 1.9.2 (up to date; next check at 2017/03/02 17:21:24)

        Service: router
       Protocol: weave 1..2
           Name: a2:79:76:d0:ba:bd(core-02)
     Encryption: enabled
  PeerDiscovery: enabled
        Targets: 1
    Connections: 1 (1 established)
          Peers: 2 (with 2 established connections)
 TrustedSubnets: none

        Service: dns
         Domain: kontena.local.
       Upstream: 10.0.2.3
            TTL: 1
        Entries: 5

        Service: plugin
     DriverName: weave

core@core-02 ~ $ docker run --rm --net host --pid host --privileged -v /var/run/docker.sock:/var/run/docker.sock weaveworks/weaveexec:1.9.2 status peers
a2:79:76:d0:ba:bd(core-02)
   -> 192.168.66.101:6783   c6:3f:e0:48:25:8d(core-01)            established
c6:3f:e0:48:25:8d(core-01)
   <- 192.168.66.102:40203  a2:79:76:d0:ba:bd(core-02)            established
core@core-02 ~ $ docker run --rm --net host --pid host --privileged -v /var/run/docker.sock:/var/run/docker.sock weaveworks/weaveexec:1.9.2 status connections
-> 192.168.66.101:6783   established encrypted   fastdp c6:3f:e0:48:25:8d(core-01) mtu=1376
```

**Comment by SpComb** [comment]
All host node traffic is now IPSec ESP per default, and no longer UDP port 6784. This needs some docs updates:

* https://github.com/kontena/kontena/tree/master/docs/getting-started/installing#on-nodes
* https://github.com/kontena/kontena/blob/master/docs/core-concepts/networking.md#node-infrastructure-services

**Comment by SpComb** [comment]
So there are a total of three different network-level protocols that may be used between agents now:

* UDP port 6783 (`weaver` userspace sleeve)
* UDP port 6784 (kernel + VXLAN)
* ESP (kernel IPSec encrypted UDP port 6784 + VXLAN)

**Comment by SpComb** [comment]
See the weave docs for additional details to consider: https://www.weave.works/documentation/net-latest-using-weave/net-latest-fastdp/

> Each encrypted dataplane packet is encapsulated into ESP, thus in some networks a firewall rule for allowing ESP traffic needs to be installed. E.g. Google Cloud Platform denies ESP packets by default.

**Comment by jakolehm** [comment]
I did some benchmarks while ago (when Weave Net 1.9.0) was released.. Ipsec is much faster than sleeve but it's still good idea to use trusted-subnets wherever it's possible (because that has almost native performance).

**Comment by brb** [comment]
@jakolehm would you mind sharing the numbers?

**Comment by jakolehm** [comment]
@brb see here: https://gist.github.com/jakolehm/d3ad43baffc2ce92795935fc306b3049

I did run those benchmarks multiple times and numbers were about the same until I upgraded droplets to 8GB.. then performance actually dropped but I think the reason was that there were more noisy neighbours. 

Maybe we should run bechmarks in bare metal (Packet.net) to avoid these kind of problems.

File: docs/core-concepts/networking.md [hunk 1]
```diff
@@ -125,6 +125,7 @@ Each host Node runs a number of infrastructure services as Docker containers, us
 | Weave Net | TCP      | 6783 | `*`                       | Weave Net Control
 | Weave Net | UDP      | 6783 | `*`                       | Weave Net Data (`sleeve`)
 | Weave Net | UDP      | 6784 | `*`                       | Weave Net Data (`fastdp`)
+| Weave Net | ESP      | 6784 | `*`                       | Weave Net Data (`IPSec fastdp`)
```
Comment (SpComb) [inline docs/core-concepts/networking.md]: I'd mark the Protocol as `ESP + UDP` or `ESP (UDP)` or somesuch... ESP + a port number doesn't make sense :)

File: docs/core-concepts/networking.md [hunk 2]
```diff
@@ -137,11 +138,25 @@ The overlay network is established using the network addresses of the peer Nodes
...
+
+[Weave's Fast Datapath](https://www.weave.works/docs/net/latest/using-weave/fastdp/) can be used for traffic between Nodes within the Kontena Grid's [Trusted Subnets](https://kontena.io/docs/using-kontena/grids#grid-trusted-subnets).
 Using Trusted Subnets and Weave's Fast Datapath provides [improved performance](https://www.weave.works/weave-docker-networking-performance-fast-data-path/). The cost of this method is a lack of data plane encryption between Nodes.
 
+### Encrypted Datapath
+
+If no `trusted subnets` have been configured in a grid, the overlay network creates IPSec encrypted Fast Datapath connections between the peers.
+
+**NOTE:**
+Each encrypted dataplane packet is encapsulated into ESP, thus in some networks a firewall rule for allowing ESP traffic needs to be installed. E.g. Google Cloud Platform denies ESP packets by default. When upgrading from a version earlier than 1.2.0, if your host's network interface has a limit on packet size (the "MTU") smaller than 1496 bytes, you should reboot after upgrading to ensure encrypted fast datapath can work. For instance this applies to Google Cloud Platform, but is not necessary on AWS.
```
Comment (SpComb) [inline docs/core-concepts/networking.md]: The 1.2.0 upgrade reboot instructions should be in the 1.2.0 release notes, and I would leave it out of the docs, it's not where people are going to notice it or go looking for it.

File: docs/getting-started/installing/README.md [hunk 3]
```diff
@@ -43,4 +43,5 @@ To operate properly Kontena needs only a few ports opened in firewalls. The prov
 
 **Between nodes:**
 
+* [ESP](https://tools.ietf.org/html/rfc2406) traffic needs to be allowed.
```
Comment (SpComb) [inline docs/getting-started/installing/README.md]: IPSec (ESP). Just "ESP" by itself is kinda obscure.

File: docs/core-concepts/networking.md [hunk 4]
```diff
@@ -137,11 +138,25 @@ The overlay network is established using the network addresses of the peer Nodes
...
+
+### Sleeve
+
+If Fast Datapath connections cannot be established, Weave automatically makes a fallback using [Weave's encrypted `sleeve` tunnels](https://www.weave.works/docs/net/latest/using-weave/security-untrusted-networks/). Sleeve tunnels use user space encryption thus the sleeve connections are quite a bit slower than Fast Datapath connections.
```
Comment (SpComb) [inline docs/core-concepts/networking.md]: The linked "Securing Connections Across Untrusted Networks" seems a bit out of date, the [Using Fast Datapath](https://www.weave.works/docs/net/latest/using-weave/fastdp/) page seems more relevant to this:

> When Weave Net cannot use the fast data path between two hosts, it falls back to a slower packet forwarding approach called sleeve. Selecting the fastest forwarding approach is automatic, and is determined on a connection-by-connection basis. For example, a Weave network spanning two data centers might use fast data path within the data centers, but not for the more constrained network link between them.

File: docs/core-concepts/networking.md [hunk 5]
```diff
@@ -137,11 +138,25 @@ The overlay network is established using the network addresses of the peer Nodes
...
 
-Alternatively, [Weave's Fast Datapath](https://www.weave.works/docs/net/latest/using-weave/fastdp/) can be used for traffic between Nodes within the Kontena Grid's [Trusted Subnets](https://kontena.io/docs/using-kontena/grids#grid-trusted-subnets).
+### Fast Datapath
+
+[Weave's Fast Datapath](https://www.weave.works/docs/net/latest/using-weave/fastdp/) can be used for traffic between Nodes within the Kontena Grid's [Trusted Subnets](https://kontena.io/docs/using-kontena/grids#grid-trusted-subnets).
 Using Trusted Subnets and Weave's Fast Datapath provides [improved performance](https://www.weave.works/weave-docker-networking-performance-fast-data-path/). The cost of this method is a lack of data plane encryption between Nodes.
 
+### Encrypted Datapath
+
+If no `trusted subnets` have been configured in a grid, the overlay network creates IPSec encrypted Fast Datapath connections between the peers.
```
Comment (SpComb) [inline docs/core-concepts/networking.md]: > In the default configuration without any Grid [Trusted Subnets](https://kontena.io/docs/using-kontena/grids#grid-trusted-subnets) configured, the overlay network wlll use IPSec encrypted Fast Datapath connections between peers.

File: docs/core-concepts/networking.md (hunk 4 above)
Comment (SpComb) [inline docs/core-concepts/networking.md]: > If firewall or MTU restrictions prevent the use of IPSec encrypted `fastdp` connections, Weave will automatically fall back to `sleeve` tunnels for the connection. The userspace `sleeve` tunnels will have slower performance than the `fastdp` connections.

File: docs/core-concepts/networking.md (hunk 2 above)
Comment (SpComb) [inline docs/core-concepts/networking.md]: Hmm... do we have something to link to for the GCP stuff? Maybe:

> The Google Cloud Platform [default firewall rules](https://cloud.google.com/compute/docs/networking#firewall_rules) do not allow IPSec ESP connections, and you must configure the firewall rules in order to use encrypted Weave `fastdp`.

Bonus points for actually testing Kontena 1.2.0 on GCP and including specific instructions here... separate PR :D

File: docs/core-concepts/networking.md [hunk 6]
```diff
@@ -137,11 +138,25 @@ The overlay network is established using the network addresses of the peer Nodes
 The overlay network mesh uses the Private network address of a peer node within the same Region; otherwise, the Public network address is used.
 Once the overlay network is started, the host Node's overlay network address is configured using [`weave expose`](https://www.weave.works/docs/net/latest/using-weave/host-network-integration/).
 
-The overlay network is powered by Weave Net, using [Weave's encrypted `sleeve` tunnels](https://www.weave.works/docs/net/latest/using-weave/security-untrusted-networks/) to form a flat Layer 2 network spanning all Grid Nodes and connected Containers.
+The overlay network, powered by Weave Net, is created to form a flat Layer 2 network spanning all Grid Nodes and connected Containers. Overlay network operates at three different modes depending on the environment and grid configuration, see below for their descriptions. **Selecting the fastest forwarding approach is automatic, and is determined on a connection-by-connection basis.**
```
Comment (SpComb) [inline docs/core-concepts/networking.md]: > The overlay network, powered by Weave Net, forms a flat ...

> The overlay network operates in three different modes ...

**Comment by SpComb** [comment]
https://github.com/weaveworks/weave/releases/tag/v1.9.3

**Pull request merged**